unsize/src/lib.rs

```diff
@@ -393,3 +397,47 @@ unsafe fn unsize_with<T, U: ?Sized>(
 /// let _ = ptr.as_sized_ptr();
 /// ```
 extern {}
+
+macro_rules! unbounded_derives {(
+    #![helper = $StructName_:ident]
+    #[doc = $doc:expr]
+    $( #[$attr:meta] )*
+    $pub:vis struct $StructName:ident [$($generics:tt)*] {
+        $(
+            $( #[$field_attr:meta] )*
+            $field_pub:vis
+            $field_name:ident : $FieldTy:ty
+        ),* $(,)?
+    }
+) => (
+    #[cfg(doc)]
+    #[doc = $doc]
+    $( #[$attr] )*
+    $pub struct $StructName< $($generics)* > {
+        $(
+            $( #[$field_attr] )*
+            $field_pub
+            $field_name: $FieldTy,
+        )*
+    }
+
+    #[cfg(not(doc))]
+    $( #[$attr] )*
+    #[allow(missing_docs, nonstandard_style)]
+    $pub struct $StructName_< $($field_name),* > {
+        $(
+            $( #[$field_attr] )*
+            $field_pub
+            $field_name: $field_name
+        ),*
+    }
+
+    #[cfg(not(doc))]
+    #[allow(type_alias_bounds)]
+    #[doc = $doc]
+    $pub type $StructName<$($generics)*> = $StructName_< $($FieldTy),* >;
+)}
+pub(crate) use unbounded_derives;
```

Comment:
I find the macro to be _super_ confusing. For this one usage in the crate I'd rather see custom impls than that. Anything where the doc deviates from the actual implementation is fishy enough but not being able to read the field types.. I don't know.

Reply:
Fair enough; this is more of an itch I've had lately with how "stupid" the `#[derive]`s are, and how custom impls, be it through a custom derive, or by manually writing them, are suboptimal since they either lead to code duplication when hand-rolling them, or to a proc-macro dependency when using a helper crate, and for the case of `Eq`, it leads to losing `StructuralEq` on stable Rust in both cases.

This approach of mine does have the advantage of not having these drawbacks, and yielding no functionally different code between `cfg(doc)` and `cfg(not(doc))`. It's main drawback is actually being unprecedented, and thus, poorly readable indeed.

This is thus a bit of a vicious cycle kind of argument, but I am obviously not gonna _impose_ my odd style in someone else's crate 😉 I'll use manual impls then 👍

Reply:
Regarding `StructuralEq`: The potential `const` usage is currently blocked on a number of other issues, such as not being allowed to use raw pointer types, function pointers, etc. I wouldn't worry about losing the property.